### ISSUE TITLE
C++11 Move Operators / Constructors

### ISSUE THREAD

**Issue by friedmud** [libMesh/libmesh#994]
I just wanted to get the conversation started about how to get C++11 move operators / constructors into libMesh.

In particular I'd like to enable them for `TypeVector` (and its derived classes).

We had some discussion about this here: #849

It seems to me that the default move operator / constructor would most likely do the right thing for `TypeVector`.  So just adding these two lines should get the job done:

```C++
TypeVector(TypeVector &&) = default;
TypeVector & TypeVector::operator= (TypeVector &&) = default;
```

Should we just wrap that up in a `LIBMESH_HAVE_CXX11`?

**Comment by jwpeterson** [comment]
> On Jun 4, 2016, at 12:50 PM, Derek Gaston <notifications@github.com> wrote:
> 
> I just wanted to get the conversation started about how to get C++11 move operators / constructors into libMesh.
> 
> In particular I'd like to enable them for TypeVector (and its derived classes).
> 
> We had some discussion about this here: #849
> 
> It seems to me that the default move operator / constructor would most likely do the right thing for TypeVector. So just adding these two lines should get the job done:
> 
> TypeVector(TypeVector &&) = default;
> TypeVector & TypeVector::operator= (TypeVector &&) = default;
> Should we just wrap that up in a LIBMESH_HAVE_CXX11?
> 

Yeah, LIBMESH_HAVE_CXX11_RVALUE_REFERENCES would probably be slightly more precise.

**Comment by friedmud** [comment]
Thanks @jwpeterson I'll give it a shot...

**Comment by roystgnr** [comment]
This should probably crossreference #655.

Speaking of move operations and the #655 checklist - it looks like (e.g. https://github.com/idaholab/moose/pull/7201#discussion_r67607447 ) you're starting to use std::move() in Moose and apps?  Any notable benchmark improvments?  If so then it's probably time to add libmesh_move()

**Comment by friedmud** [comment]
@roystgnr I don't believe we've done any benchmarks yet... but the new code is SO much better.  For instance, @permcody has transitioned to storing large objects in `std::vector`s instead of storing pointers... and using move operators to ensure it's all done efficiently.  This makes for really nice looking code.  As an example look here: https://github.com/idaholab/moose/pull/7161/files

`move` is not "just" about efficiency... it also enables different code design by freeing us from the shackles of copying all over the place in C++.

**Comment by friedmud** [comment]
To be clear: I'm bringing that up because a `libmesh_move()`, by itself, would NOT allow for changing of the algorithm / storage... because storing objects instead of pointers in `std::vector`s would be INCREDIBLY slow in C++98.

To really realize the potential of `move` we need to transition the library to C++11...

**Comment by permcody** [comment]
Derek beat me to the punch. I don't expect much in terms of speed
improvements but you can certainly write more natural code (less pointer
indirection).

**Comment by jwpeterson** [comment]
> because storing objects instead of pointers in std::vectors would be INCREDIBLY slow in C++98.

I looked at implementing the `libmesh_move()` item on this ticket, but it's one of those things that doesn't really make sense without adding the appropriate move-constructor and move-assignment operators as well, *and* changing the code to use objects instead of pointers.  I figured it was something that would be best to implement after C++11 was officially required in libmesh.

**Comment by roystgnr** [comment]
And now it looks like we're ready to officially require C++11 in libMesh.  We'll want to put out one more official release first to keep my users happy, and I'll ping the libmesh_users mailing list to see if there's anyone else who might be unhappy, but I think libMesh 1.3 (or libMesh 2.0?  This is probably big enough to justify a major version number increment) can be C++11-only.
